File: generated_code/GPT4/Python/058_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = 'key', 'left', 'right', 'parent'

        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    def _rotate_left(self, x):
        y = x.right
        if y is None:
            return
        x.right = y.left
        if y.left:
            y.left.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.left = x
        x.parent = y

    def _rotate_right(self, x):
        y = x.left
        if y is None:
            return
        x.left = y.right
        if y.right:
            y.right.parent = x
        y.parent = x.parent
        if x.parent is None:
            self.root = y
        else:
            if x == x.parent.left:
                x.parent.left = y
            else:
                x.parent.right = y
        y.right = x
        x.parent = y

    def _splay(self, x):
        while x.parent:
            if x.parent.parent is None:
                # Zig
                if x == x.parent.left:
                    self._rotate_right(x.parent)
                else:
                    self._rotate_left(x.parent)
            else:
                if x == x.parent.left and x.parent == x.parent.parent.left:
                    # Zig-Zig (left left)
                    self._rotate_right(x.parent.parent)
                    self._rotate_right(x.parent)
                elif x == x.parent.right and x.parent == x.parent.parent.right:
                    # Zig-Zig (right right)
                    self._rotate_left(x.parent.parent)
                    self._rotate_left(x.parent)
                elif x == x.parent.right and x.parent == x.parent.parent.left:
                    # Zig-Zag (left right)
                    self._rotate_left(x.parent)
                    self._rotate_right(x.parent)
                else:
                    # Zig-Zag (right left)
                    self._rotate_right(x.parent)
                    self._rotate_left(x.parent)

    def search(self, key):
        """
        Returns True if key exists in the set, else False.
        After search, the accessed node (found or last accessed) is splayed to root.
        """
        x = self.root
        prev = None
        while x:
            prev = x
            if key == x.key:
                self._splay(x)
                return True
            elif key < x.key:
                x = x.left
            else:
                x = x.right
        if prev:
            self._splay(prev)
        return False

    def insert(self, key):
        """
        Inserts key into the set if not already present.
        """
        if not self.root:
            self.root = self.Node(key)
            return
        x = self.root
        while True:
            if key == x.key:
                self._splay(x)
                return
            elif key < x.key:
                if x.left:
                    x = x.left
                else:
                    x.left = self.Node(key, parent=x)
                    self._splay(x.left)
                    return
            else:
                if x.right:
                    x = x.right
                else:
                    x.right = self.Node(key, parent=x)
                    self._splay(x.right)
                    return

    def _subtree_minimum(self, x):
        while x.left:
            x = x.left
        return x

    def _replace(self, u, v):
        if u.parent is None:
            self.root = v
        else:
            if u == u.parent.left:
                u.parent.left = v
            else:
                u.parent.right = v
        if v:
            v.parent = u.parent

    def delete(self, key):
        """
        Deletes key from the set if present.
        """
        x = self.root
        while x:
            if key == x.key:
                self._splay(x)
                break
            elif key < x.key:
                x = x.left
            else:
                x = x.right
        else:
            # Not found, splay last accessed node if any
            if self.root:
                # Find closest
                self.search(key)
            return  # Key not found

        # Now x is at root
        if x.left is None:
            self._replace(x, x.right)
        elif x.right is None:
            self._replace(x, x.left)
        else:
            # Find min in right subtree
            y = self._subtree_minimum(x.right)
            if y.parent != x:
                self._replace(y, y.right)
                y.right = x.right
                y.right.parent = y
            self._replace(x, y)
            y.left = x.left
            y.left.parent = y
        # No need to splay after deletion

    # Optional: for debugging and testing
    def __contains__(self, key):
        return self.search(key)

    def __str__(self):
        def print_tree(node, indent=""):
            if not node:
                return ""
            result = ""
            result += print_tree(node.right, indent + "   ")
            result += f"{indent}{node.key}\n"
            result += print_tree(node.left, indent + "   ")
            return result
        return print_tree(self.root)

File: generated_code/GPT4/Java/058_SplayTree.java
public class SplayTree {
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key, Node parent) {
            this.key = key;
            this.parent = parent;
        }
    }

    private Node root;

    /**
     * Inserts the key into the set.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key, null);
            return;
        }
        Node z = root, p = null;
        while (z != null) {
            p = z;
            if (key < z.key)
                z = z.left;
            else if (key > z.key)
                z = z.right;
            else {
                splay(z);
                return; // Key already exists, do not insert duplicates
            }
        }
        Node n = new Node(key, p);
        if (key < p.key)
            p.left = n;
        else
            p.right = n;
        splay(n);
    }

    /**
     * Deletes the key from the set if it exists.
     */
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null)
            return;
        splay(node);

        if (node.left == null) {
            replace(node, node.right);
        } else if (node.right == null) {
            replace(node, node.left);
        } else {
            Node min = subtreeMin(node.right);
            if (min.parent != node) {
                replace(min, min.right);
                min.right = node.right;
                if (min.right != null)
                    min.right.parent = min;
            }
            replace(node, min);
            min.left = node.left;
            if (min.left != null)
                min.left.parent = min;
        }
        node.left = node.right = node.parent = null; // Help GC
    }

    /**
     * Searches for the key in the set.
     * If found, returns the key (as Integer), else returns null.
     * The accessed node (or last accessed) is splayed to the root if any access occurred.
     */
    public Integer search(int key) {
        Node node = root;
        Node last = null;
        while (node != null) {
            last = node;
            if (key < node.key)
                node = node.left;
            else if (key > node.key)
                node = node.right;
            else {
                splay(node);
                return node.key;
            }
        }
        if (last != null)
            splay(last); // Splay the last accessed node
        return null;
    }

    // Helper: Find node by key (no splaying)
    private Node findNode(int key) {
        Node node = root;
        while (node != null) {
            if (key < node.key)
                node = node.left;
            else if (key > node.key)
                node = node.right;
            else
                return node;
        }
        return null;
    }

    // Helper: Replace u with v in the parent
    private void replace(Node u, Node v) {
        if (u.parent == null) {
            root = v;
        } else if (u == u.parent.left) {
            u.parent.left = v;
        } else {
            u.parent.right = v;
        }
        if (v != null)
            v.parent = u.parent;
    }

    // Helper: Splay operation
    private void splay(Node x) {
        if (x == null)
            return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig
                if (x == x.parent.left)
                    rotateRight(x.parent);
                else
                    rotateLeft(x.parent);
            } else if (x == x.parent.left && x.parent == x.parent.parent.left) {
                // Zig-zig
                rotateRight(x.parent.parent);
                rotateRight(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.right) {
                // Zig-zig
                rotateLeft(x.parent.parent);
                rotateLeft(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.left) {
                // Zig-zag
                rotateLeft(x.parent);
                rotateRight(x.parent);
            } else {
                // Zig-zag
                rotateRight(x.parent);
                rotateLeft(x.parent);
            }
        }
    }

    // Left rotation at x
    private void rotateLeft(Node x) {
        Node y = x.right;
        if (y == null)
            return;
        x.right = y.left;
        if (y.left != null)
            y.left.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.left = x;
        x.parent = y;
    }

    // Right rotation at x
    private void rotateRight(Node x) {
        Node y = x.left;
        if (y == null)
            return;
        x.left = y.right;
        if (y.right != null)
            y.right.parent = x;
        y.parent = x.parent;
        if (x.parent == null)
            root = y;
        else if (x == x.parent.left)
            x.parent.left = y;
        else
            x.parent.right = y;
        y.right = x;
        x.parent = y;
    }

    // Helper: Find subtree minimum
    private Node subtreeMin(Node node) {
        while (node.left != null)
            node = node.left;
        return node;
    }

    // Optional: For debug
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        print(root, sb, "", "");
        return sb.toString();
    }

    private void print(Node n, StringBuilder sb, String prefix, String childrenPrefix) {
        if (n == null) return;
        sb.append(prefix);
        sb.append(n.key);
        sb.append('\n');
        if (n.right != null) {
            print(n.right, sb, childrenPrefix + "├── ", childrenPrefix + "│   ");
        }
        if (n.left != null) {
            print(n.left, sb, childrenPrefix + "└── ", childrenPrefix + "    ");
        }
    }
}